src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+	err = readString(reader, &param.StrDefault)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.Value)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.Indexed)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.InternalType)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &param.Components, ReadParameter)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.Unused)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.IsFlags)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadTrace(reader *bufio.Reader) (trace *types.SimpleTrace, err error) {
+	trace = &types.SimpleTrace{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &trace.BlockHash)
+	if err != nil {
+		return
+	}
+
+	err = read(&trace.BlockNumber)
+	if err != nil {
+		return
+	}
+
+	err = read(&trace.Subtraces)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &trace.TraceAddress, makeArrayItemRead(reader, readString))
+	if err != nil {
+		return
+	}
+
+	err = readHash(reader, &trace.TransactionHash)
+	if err != nil {
+		return
+	}
+
+	err = read(&trace.TransactionIndex)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &trace.TraceType)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &trace.Error)
+	if err != nil {
+		return
+	}
+
+	action, err := readTraceAction(reader)
+	if err != nil {
+		return
+	}
+	trace.Action = action
+
+	result, err := readTraceResult(reader)
+	if err != nil {
+		return
+	}
+	trace.Result = result
+
+	function, err := ReadFunction(reader)
+	if err != nil {
+		return
+	}
+	trace.ArticulatedTrace = function
+
+	return
+}
+
+func readTraceAction(reader *bufio.Reader) (action *types.SimpleTraceAction, err error) {
+	action = &types.SimpleTraceAction{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &action.SelfDestructed)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &action.Balance)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &action.CallType)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &action.From)
+	if err != nil {
+		return
+	}
+
+	err = read(&action.Gas)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &action.Init)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &action.Input)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &action.RefundAddress)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &action.To)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &action.Value)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func readTraceResult(reader *bufio.Reader) (result *types.SimpleTraceResult, err error) {
```

Comment:
Eventually, we'll move this into the auto-generated type specific file. Please add a TODO, but don't move this yet.